src/mlpack/methods/neighbor_search/knn_low_resource_main.cpp

```diff
@@ -0,0 +1,188 @@
+#include <mlpack/core/tree/cover_tree.hpp>
+#include <mlpack/core/util/cli.hpp>
+#include <mlpack/core/util/mlpack_main.hpp>
+#include <mlpack/prereqs.hpp>
+
+// #include <fstream>
+// #include <iostream>
+// #include <string>
+
+#include "neighbor_search.hpp"
+#include "ns_model.hpp"
+#include "unmap.hpp"
+
+using namespace std;
+using namespace mlpack;
+using namespace mlpack::neighbor;
+using namespace mlpack::tree;
+using namespace mlpack::metric;
+using namespace mlpack::util;
+
+// Convenience typedef.
+typedef NSModel<NearestNeighborSort> KNNModel;
+
+// Information about the program itself.
+PROGRAM_INFO(
+  "k-Nearest-Neighbors Search",
+  // Short description.
+  "An implementation of k-nearest-neighbor search using "
+  "dual-tree algorithms.  Given a set of reference points and query points, "
+  "this can find the k nearest neighbors in the reference set of each query "
+  "point using trees; trees that are built can be saved for future use.",
+  // Long description.
+  "This program will calculate the k-nearest-neighbors of a set of "
+  "points using kd-trees or cover trees (cover tree support is experimental "
+  "and may be slow). You may specify a separate set of "
+  "reference points and query points, or just a reference set which will be "
+  "used as both the reference and query set."
+  "\n\n"
+  "For example, the following command will calculate the 5 nearest neighbors "
+  "of each point in " +
+    PRINT_DATASET("input") +
+    " and store the distances "
+    "in " +
+    PRINT_DATASET("distances") + " and the neighbors in " +
+    PRINT_DATASET("neighbors") +
+    ": "
+    "\n\n" +
+    PRINT_CALL("knn",
+               "k",
+               5,
+               "reference",
+               "input",
+               "neighbors",
+               "neighbors",
+               "distances",
+               "distances") +
+    "\n\n"
+    "The output is organized such that row i and column j in the neighbors "
+    "output matrix corresponds to the index of the point in the reference set "
+    "which is the j'th nearest neighbor from the point in the query set with "
+    "index i.  Row j and column i in the distances output matrix corresponds to"
+    " the distance between those two points.",
+  SEE_ALSO("NeighborSearch tutorial (k-nearest-neighbors)",
+           "@doxygen/nstutorial.html"),
+  SEE_ALSO("Tree-independent dual-tree algorithms (pdf)",
+           "http://proceedings.mlr.press/v28/curtin13.pdf"),
+  SEE_ALSO("mlpack::neighbor::NeighborSearch C++ class documentation",
+           "@doxygen/classmlpack_1_1neighbor_1_1NeighborSearch.html"));
+
+ // Define our input parameters that this program will take.
+ PARAM_MATRIX_OUT("distances", "Matrix to output distances into.", "d");
+ PARAM_UMATRIX_OUT("neighbors", "Matrix to output neighbors into.", "n");
+
+ // // The option exists to load or save models.
+//PARAM_MODEL_IN(KNNModel, "input_model", "Pre-trained kNN model.", "m");
```

Comment:
If you want to avoid serialization too, try adding back in the `reference` parameter, and then constructing a KNN object like this:

```
arma::mat referenceSet = std::move(CMD::GetParam<arma::mat>("reference"));
KNN<> knn(std::move(referenceSet), DUAL_TREE_MODE);
```

I think I have the signature right there, but I'm not 100% sure, so please forgive me if it fails to compile like that. :smile: